File: src/main/java/no/fint/FintMimeMessage.java
package no.fint;

import javax.mail.Message;
import javax.mail.MessagingException;
import javax.mail.PasswordAuthentication;
import javax.mail.Session;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.util.Properties;

public class FintMimeMessage {

    private final MimeMessage mimeMessage;

    public FintMimeMessage(Config config, String content, String subject) throws MessagingException {
        mimeMessage = new MimeMessage(createSession(config));

        mimeMessage.setFrom(new InternetAddress(config.getSender()));

        for (String recipient : config.getRecipients()) { addRecipient(recipient); }
        mimeMessage.setSubject(subject);
        mimeMessage.setText(content, "UTF-8", "html");
    }

    private void addRecipient(String recipient) throws MessagingException {
        mimeMessage.addRecipient(Message.RecipientType.TO,new InternetAddress(recipient));
    }

    private Session createSession(Config config) {
        Properties props = new Properties();
        props.put("mail.smtp.auth", "true");
        props.put("mail.smtp.starttls.enable", "true");
        props.put("mail.smtp.host", config.getSmtpServer());
        props.put("mail.smtp.port", config.getSmtpPort());

        return Session.getInstance(props,
                new javax.mail.Authenticator() {
                    protected PasswordAuthentication getPasswordAuthentication() {
                        return new PasswordAuthentication(config.getSmtpUsername(), config.getSmtpPassword());
                    }
                });
    }

    public MimeMessage getMail() {
        return mimeMessage;
    }
}

File: src/main/kotlin/no/fint/mailing/FintMimeMessage.kt
package no.fint.mailing

import no.fint.Config
import java.util.Properties
import javax.mail.Authenticator
import javax.mail.Message
import javax.mail.PasswordAuthentication
import javax.mail.Session
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeMessage

class FintMimeMessage(
    config: Config,
    content: String,
    subject: String,
) {
    val mail: MimeMessage =
        MimeMessage(createSession(config)).apply {
            setFrom(InternetAddress(config.sender))
            config.recipients.forEach { recipient ->
                addRecipient(Message.RecipientType.TO, InternetAddress(recipient))
            }
            setSubject(subject)
            setText(content, "UTF-8", "html")
        }

    private fun createSession(config: Config): Session {
        val props =
            Properties().apply {
                put("mail.smtp.auth", "true")
                put("mail.smtp.starttls.enable", "true")
                put("mail.smtp.host", config.smtpServer)
                put("mail.smtp.port", config.smtpPort)
            }

        return Session.getInstance(
            props,
            object : Authenticator() {
                override fun getPasswordAuthentication() = PasswordAuthentication(config.smtpUsername, config.smtpPassword)
            },
        )
    }
}
